reasoner/planner/ConjunctionStreamPlan.java

```diff
@@ -0,0 +1,337 @@
+            Set<Retrievable> nextSuffixVariables = new HashSet<>(subConjunction.get(subConjunction.size() - 1).retrieves());
+            int r;
+            for (r = subConjunction.size() - 2; r > 0; r--) { // r > 0 because we need atleast one thing in our left
+                nextSuffixVariables.addAll(subConjunction.get(r).retrieves());
+                Set<Retrievable> subtreeBounds = VariableSets.intersection(boundsAt.get(r), nextSuffixVariables);
+                Set<Retrievable> candidateFirstChildBounds = VariableSets.intersection(subConjunction.get(r).retrieves(), boundsAt.get(r));
+
+                if (!VariableSets.difference(subtreeBounds, candidateFirstChildBounds).isEmpty()) {
+                    break;
+                }
+            }
+            return r;
+        }
+
+        private ConjunctionStreamPlan flatten(ConjunctionStreamPlan conjunctionStreamPlan) {
+            if (conjunctionStreamPlan.isResolvable()) {
+                return conjunctionStreamPlan;
+            } else {
+                CompoundStreamPlan compoundStreamPlan = conjunctionStreamPlan.asCompoundStreamPlan();
+                assert compoundStreamPlan.size() == 2;
+                List<ConjunctionStreamPlan> subPlans = new ArrayList<>();
+                for (int i = 0; i < 2; i++) {
+                    ConjunctionStreamPlan unflattenedChild = compoundStreamPlan.ithChild(i);
+
+                    if (canFlattenInto(compoundStreamPlan, unflattenedChild)) {
+                        subPlans.addAll(flatten(unflattenedChild).asCompoundStreamPlan().subPlans);
+                    } else {
+                        subPlans.add(flatten(unflattenedChild));
+                    }
+                }
+                return new CompoundStreamPlan(subPlans, compoundStreamPlan.identifierVariables(), compoundStreamPlan.extendOutputWithVariables(), compoundStreamPlan.outputVariables());
+            }
+        }
+
+        private boolean canFlattenInto(ConjunctionStreamPlan.CompoundStreamPlan conjunctionStreamPlan, ConjunctionStreamPlan unflattenedChild) {
+            return unflattenedChild.isCompoundStream() &&
+                    isExclusiveReader(conjunctionStreamPlan.asCompoundStreamPlan(), unflattenedChild.asCompoundStreamPlan(), inputBounds) &&
+                    rightChildBoundsSatisfied(conjunctionStreamPlan.asCompoundStreamPlan(), unflattenedChild.asCompoundStreamPlan());
+        }
+
+        private static boolean rightChildBoundsSatisfied(ConjunctionStreamPlan.CompoundStreamPlan conjunctionStreamPlan, ConjunctionStreamPlan.CompoundStreamPlan unflattenedNextPlan) {
+            return VariableSets.difference(
+                            unflattenedNextPlan.ithChild(1).identifierVariables,
+                            VariableSets.union(conjunctionStreamPlan.identifierVariables, unflattenedNextPlan.ithChild(0).outputVariables))
+                    .isEmpty();
+        }
+
+        private static class VariableSets {
```

Comment:
Holds the logic for determining the sets of input, join and output variables for the left and right subtrees.